airbyte-scheduler/src/test/java/io/airbyte/scheduler/JobTrackerTest.java

```diff
@@ -0,0 +1,204 @@
+    testAsynchronous(ConfigType.SYNC);
+  }
+
+  @Test
+  void testTrackReset() throws ConfigNotFoundException, IOException, JsonValidationException {
+    testAsynchronous(ConfigType.RESET_CONNECTION);
+  }
+
+  void testAsynchronous(ConfigType configType) throws ConfigNotFoundException, IOException, JsonValidationException {
+    // for sync the job id is a long not a uuid.
+    final long jobId = 10L;
+
+    final ImmutableMap<String, Object> metadata = ImmutableMap.<String, Object>builder()
+        .put("job_type", configType)
+        .put("job_id", String.valueOf(jobId))
+        .put("attempt_id", 700)
+        .put("connection_id", CONNECTION_ID)
+        .put("connector_source", SOURCE_DEF_NAME)
+        .put("connector_source_definition_id", UUID1)
+        .put("connector_destination", DESTINATION_DEF_NAME)
+        .put("connector_destination_definition_id", UUID2)
+        .build();
+
+    when(configRepository.getSourceDefinitionFromConnection(CONNECTION_ID))
+        .thenReturn(new StandardSourceDefinition().withSourceDefinitionId(UUID1).withName(SOURCE_DEF_NAME));
+    when(configRepository.getDestinationDefinitionFromConnection(CONNECTION_ID))
+        .thenReturn(new StandardDestinationDefinition().withDestinationDefinitionId(UUID2).withName(DESTINATION_DEF_NAME));
+
+    when(configRepository.getStandardSourceDefinition(UUID1))
+        .thenReturn(new StandardSourceDefinition().withSourceDefinitionId(UUID1).withName(SOURCE_DEF_NAME));
+    when(configRepository.getStandardDestinationDefinition(UUID2))
+        .thenReturn(new StandardDestinationDefinition().withDestinationDefinitionId(UUID2).withName(DESTINATION_DEF_NAME));
+
+    final Job job = mock(Job.class);
+    when(job.getId()).thenReturn(jobId);
+    when(job.getConfigType()).thenReturn(configType);
+    when(job.getScope()).thenReturn(CONNECTION_ID.toString());
+    when(job.getAttemptsCount()).thenReturn(700);
+
+    // test when frequency is manual.
+    when(configRepository.getStandardSyncSchedule(CONNECTION_ID)).thenReturn(new StandardSyncSchedule().withManual(true));
+    final Map<String, Object> manualMetadata = MoreMaps.merge(metadata, ImmutableMap.of("frequency", "manual"));
+    assertCorrectMessageForEachState((jobState) -> jobTracker.trackSync(job, jobState), manualMetadata);
+
+    // test when frequency is scheduled.
+    when(configRepository.getStandardSyncSchedule(CONNECTION_ID))
+        .thenReturn(new StandardSyncSchedule().withManual(false).withSchedule(new Schedule().withUnits(1L).withTimeUnit(TimeUnit.MINUTES)));
+    final Map<String, Object> scheduledMetadata = MoreMaps.merge(metadata, ImmutableMap.of("frequency", "1 min"));
+    assertCorrectMessageForEachState((jobState) -> jobTracker.trackSync(job, jobState), scheduledMetadata);
+  }
+
+  private void assertCorrectMessageForEachState(Consumer<JobState> jobStateConsumer, Map<String, Object> metadata) {
+    jobStateConsumer.accept(JobState.STARTED);
```

Comment:
@cgardens took me a while to understand what this is doing - once I did, I realised this use of the consumer is top notch!

Reply:
thanks! added a comment to make this clearer.